main/src/cgeo/geocaching/models/CalcState.java

```diff
@@ -0,0 +1,124 @@
+package cgeo.geocaching.models;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+
+import cgeo.geocaching.settings.Settings;
+import cgeo.geocaching.ui.CalculateButton;
+import cgeo.geocaching.ui.CalculatorVariable;
+
+/**
+ * This class is designed to capture the current state of the coordinate calculator such that it can be preserver for latter use.
+ **
+ * All the relevant information is in a serializable form such that it can be stored as a bundle in waypoint's 'ContentValues'.
+ **/
+
+public class CalcState implements Serializable {
+    public static final char ERROR_CHAR = '#';
+    public static final String ERROR_STRING = "???";
+
+    public Settings.CoordInputFormatEnum format;
+    public String plainLat, plainLon;
+    public char latHemisphere, lonHemisphere;
+    public List<CalculateButton.ButtonData> buttons;
+    public List<CalculatorVariable.VariableData> equations, freeVariables;
+    public String notes; // Note, we have to use a String rather than an Editable as Editable's can't be serialized
+
+    public CalcState(Settings.CoordInputFormatEnum theFormat,
+                     String theLat,
+                     String theLon,
+                     char latHem,
+                     char lonHem,
+                     List<CalculateButton.ButtonData> theButtons,
+                     List<CalculatorVariable.VariableData> theEquations,
+                     List<CalculatorVariable.VariableData> theFreeVariables,
+                     String theNotes)
+    {
+        format = theFormat;
+        plainLat = theLat;
+        plainLon = theLon;
+        latHemisphere = latHem;
+        lonHemisphere = lonHem;
+        buttons = theButtons;
+        equations = theEquations;
+        freeVariables = theFreeVariables;
+        notes = theNotes;
+    }
+
+    public CalcState(JSONObject json) {
+        JSONArray array;
+        int arrayLength;
+
+        format = Settings.CoordInputFormatEnum.values()[json.optInt("format", 2)];
+        plainLat = json.optString("plainLat");
+        plainLon = json.optString("plainLon");
+        latHemisphere = (char) json.optInt("latHemisphere", ERROR_CHAR);
+        lonHemisphere = (char) json.optInt("lonHemisphere", ERROR_CHAR);
+
+        array = json.optJSONArray("buttons");
```

Comment:
Please use subblocks so that you don't reuse the variables, this is prone to error:

``` java
lonHemisphere = (char) json.optInt("lonHemisphere", ERROR_CHAR);

{
    final JSONArray array = json.optJSONArray("buttons");
    final int arrayLength = array != null ? array.length() : 0;
    …
}

{
    final JSONArray array = json.optJSONArray("equations");
    final int arrayLength = array != null ? array.length() : 0;
}
```